v2/tools/generator/internal/codegen/pipeline/fix_oneof_types.go

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright (c) Microsoft Corporation.
+ * Licensed under the MIT license.
+ */
+
+package pipeline
+
+import (
+	"context"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/Azure/azure-service-operator/v2/tools/generator/internal/astmodel"
+)
+
+// MakeOneOfDiscriminantRequiredStageId is the unique identifier for this pipeline stage
+const MakeOneOfDiscriminantRequiredStageId = "makeOneOfDiscriminantRequired"
+
+// MakeOneOfDiscriminantRequired walks the type graph and builds new types for communicating
+// with ARM
+func MakeOneOfDiscriminantRequired() Stage {
```

Comment:
Maybe this file should be called `make_one+_of_discriminant_required.go`.

Though, actually, why can't you fix this right where the discriminant is created? Why does this need a new stage?

Reply:
Will doublecheck but I think the issue was identifying it as a discriminant there.

Reply:
Yeah, so the issue is that we get a type like this:
```
    "NumberInAdvancedFilter": {
      "type": "object",
      "properties": {
        "operatorType": {
          "type": "string",
          "enum": [
            "NumberIn"
          ]
        },
        "values": {
          "oneOf": [
            {
              "type": "array",
              "items": {
                "type": "number"
              }
            },
            {
              "$ref": "https://schema.management.azure.com/schemas/common/definitions.json#/definitions/expression"
            }
          ],
          "description": "The set of filter values."
        }
      },
      "required": [
        "operatorType"
      ],
      "description": "NumberIn Advanced Filter."
    },
```

There's no field that says "This is the discriminant". The way to determine discriminant is by looking at all of the types and finding the one property they have in common (which in each case should be an enum with a single allowed value).
If we had a type that had 2 enums that satisfied those criteria we could still find the discriminant provided we have other types to compare it to, but at the point the type is being created we dont have access to those other types, so it needs to be done after the fact.

I _think_ in the Swagger there's a field that identifies the discriminant so we could maybe change this when @porges stuff lands.

Reply:
Renamed this file. Good catch.